rust/lance-encoding/src/encodings/physical/packed.rs

```diff
@@ -165,3 +176,904 @@ impl MiniBlockDecompressor for PackedStructFixedWidthMiniBlockDecompressor {
         }))
     }
 }
+
+#[derive(Debug)]
+enum VariablePackedFieldData {
+    Fixed {
+        block: FixedWidthDataBlock,
+    },
+    Variable {
+        block: VariableWidthBlock,
+        bits_per_length: u64,
+    },
+}
+
+impl VariablePackedFieldData {
+    fn append_row_bytes(&self, row_idx: usize, output: &mut Vec<u8>) -> Result<()> {
+        match self {
+            Self::Fixed { block } => {
+                let bits_per_value = block.bits_per_value;
+                if bits_per_value % 8 != 0 {
+                    return Err(Error::invalid_input(
+                        "Packed struct variable encoding requires byte-aligned fixed-width children",
+                        location!(),
+                    ));
+                }
+                let bytes_per_value = (bits_per_value / 8) as usize;
+                let start = row_idx.checked_mul(bytes_per_value).ok_or_else(|| {
+                    Error::invalid_input("Packed struct row size overflow", location!())
+                })?;
+                let end = start + bytes_per_value;
+                let data = block.data.as_ref();
+                if end > data.len() {
+                    return Err(Error::invalid_input(
+                        "Packed struct fixed child out of bounds",
+                        location!(),
+                    ));
+                }
+                output.extend_from_slice(&data[start..end]);
+                Ok(())
+            }
+            Self::Variable {
+                block,
+                bits_per_length,
+            } => {
+                if bits_per_length % 8 != 0 {
+                    return Err(Error::invalid_input(
+                        "Packed struct variable children must have byte-aligned length prefixes",
+                        location!(),
+                    ));
+                }
+                let prefix_bytes = (*bits_per_length / 8) as usize;
+                if !(prefix_bytes == 4 || prefix_bytes == 8) {
+                    return Err(Error::invalid_input(
+                        "Packed struct variable children must use 32 or 64-bit length prefixes",
+                        location!(),
+                    ));
+                }
```

Comment:
Lets make a follow-up task to investigate variable-width integer encodings for length prefixes

Reply:
Tracked in https://github.com/lancedb/lance/issues/5023